### ISSUE TITLE
Allow excluding certain areas (rects) from the balloon positioning algorithm to avoid collisions

### ISSUE THREAD

**Issue by oleq** [ckeditor/ckeditor5-ui#460]
Because the link balloon and the inline editor toolbar don't know about each other, they collide in certain situations.

![image](https://user-images.githubusercontent.com/1099479/51240251-a1ef5800-197b-11e9-9837-40b941170ae6.png)

![image](https://user-images.githubusercontent.com/1099479/51240261-a74ca280-197b-11e9-8e04-eb0853cce7fe.png)

To address this thing, we're going to need a global floating component registry probably under `editor.ui` (because not all floating components are balloons, so we cannot enclose this logic in `BalloonPanelView`) so when a component wants to position itself (e.g. link balloon), it would consider other UI elements (e.g. inline toolbar).

This could be somehow related to https://github.com/ckeditor/ckeditor5-ui/issues/181.

cc @oskarwrobel @jodator @Reinmar

**Comment by jodator** [comment]
Yeah this could be fixed by a global registry. I guess something similar that recent widget-toolbar repository introduced?

As for this particular case woudn't be enough to make content balloons appear (z-index) higher/above the editor toolbar? Just to ask what's on my mind here - I don't see this a general fix as anyway it would clutter the UI.

**Comment by oleq** [comment]
> As for this particular case woudn't be enough to make content balloons appear (z-index) higher/above the editor toolbar?

That would be a quick fix. Still, the original issue must be addressed.

**Comment by Reinmar** [comment]
DUP reported in https://github.com/ckeditor/ckeditor5/issues/1757.
